src/util/instanceMigration.tsx

```diff
@@ -0,0 +1,134 @@
+  const [type, setType] = useState<MigrationType>(
+    isClustered ? "" : "root storage pool",
+  );
+  const [target, setTarget] = useState("");
+
+  const { data: pools = [] } = useQuery({
+    queryKey: [queryKeys.storage],
+    queryFn: () => fetchStoragePools(instance.project),
+    enabled: type === "root storage pool",
+  });
+
+  const { data: members = [] } = useQuery({
+    queryKey: [queryKeys.cluster, queryKeys.members],
+    queryFn: fetchClusterMembers,
+    enabled: type === "cluster member",
+  });
+
+  const handleSuccess = (newTarget: string, instanceName: string) => {
+    let successMessage: ReactNode = "";
+    if (type === "cluster member") {
+      successMessage = (
+        <>
+          Instance <ItemName item={{ name: instanceName }} bold /> successfully
+          migrated to cluster member{" "}
+          <ItemName item={{ name: newTarget }} bold />
+        </>
+      );
+    }
+
+    if (type === "root storage pool") {
+      successMessage = (
+        <>
+          Instance <ItemName item={{ name: instanceName }} bold /> root storage
+          successfully migrated to pool{" "}
+          <ItemName item={{ name: newTarget }} bold />
+        </>
+      );
+    }
+
+    toastNotify.success(successMessage);
+  };
+
+  const notifyFailure = (e: unknown, instanceName: string) => {
+    let failureMessage = "";
+    if (type === "cluster member") {
+      failureMessage = `Cluster member migration failed for instance ${instanceName}`;
+    }
+
+    if (type === "root storage pool") {
+      failureMessage = `Root storage migration failed for instance ${instanceName}`;
+    }
+
+    instanceLoading.setFinish(instance);
+    toastNotify.failure(failureMessage, e);
+  };
+
+  const handleFailure = (msg: string, instanceName: string) => {
+    notifyFailure(new Error(msg), instanceName);
+  };
+
+  const handleFinish = () => {
+    void queryClient.invalidateQueries({
+      queryKey: [queryKeys.instances, instance.name],
+    });
+    instanceLoading.setFinish(instance);
+  };
+
+  const handleMigrate = () => {
+    instanceLoading.setLoading(instance, "Migrating");
+    const targetMember = type === "cluster member" ? target : undefined;
+    const targetPool = type === "root storage pool" ? target : undefined;
+    migrateInstance(instance.name, instance.project, targetMember, targetPool)
+      .then((operation) => {
+        eventQueue.set(
+          operation.metadata.id,
+          () => handleSuccess(target, instance.name),
+          (err) => handleFailure(err, instance.name),
+          handleFinish,
+        );
+        toastNotify.info(
+          `${capitalizeFirstLetter(type)} migration started for instance ${instance.name}`,
+        );
+        void queryClient.invalidateQueries({
+          queryKey: [queryKeys.instances, instance.name, instance.project],
+        });
+      })
+      .catch((e) => {
+        notifyFailure(e, instance.name);
+      })
+      .finally(() => {
+        close();
```

Comment:
I don't think we should call close on failure, probably only in the then block. Also I'd rename it to `onSuccess`, wdyt?

Reply:
yeah that sounds good :+1:

Reply:
The handle and on prefix is correct, I meant renaming the close prop (which doesn't fit any of the two oO) to onSuccess.

Reply:
oh my bad for misunderstanding, yes that makes sense :+1: